src/versioning.rs

```diff
@@ -0,0 +1,364 @@
+                return client.set_condition(resource, condition).await;
+            }
+            (Some(version), None) => {
+                return client
+                    .merge_patch_status(resource, &json!({ "version": version }))
+                    .await;
+            }
+            (None, Some(condition)) => {
+                return client.set_condition(resource, condition).await;
+            }
+            _ => {}
+        }
+    }
+
+    Ok(resource.clone())
+}
+
+/// Finalizes the `init_versioning`. This is required after e.g. all pods and config maps were
+/// created. It will remove the `target_version` from the status `ProductVersion` and set the
+/// condition status to false.
+///
+/// Returns the updated custom resource for further usage.
+///
+/// # Arguments
+///
+/// * `client` - The Kubernetes client.
+/// * `resource` - The cluster custom resource.
+///
+pub async fn finalize_versioning<T, S, V>(client: &Client, resource: &T) -> OperatorResult<T>
+where
+    T: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Status<S>,
+    S: Conditions + Debug + Default + Serialize + Versioned<V>,
+    V: Clone + Debug + Display + PartialEq + Serialize + Versioning,
+{
+    if let Some(status) = resource.status() {
+        if let Some(version) = status.version() {
+            if let Some(target_version) = &version.target {
+                let condition = build_versioning_condition(
+                    resource,
+                    status.conditions(),
+                    &format!(
+                        "No upgrade required [{}] is still the current_version",
+                        target_version
+                    ),
+                    "",
+                    ConditionStatus::False,
+                );
+
+                client.set_condition(resource, condition).await?;
+
+                let v = ProductVersion {
+                    current: Some(target_version.clone()),
+                    target: None,
+                };
+
+                return client
+                    .merge_patch_status(resource, &json!({ "version": v }))
+                    .await;
+            }
+        }
+    }
+
+    Ok(resource.clone())
+}
+
+/// Checks that the custom resource status (or creates the default status) and processes the contents
+/// of the `ProductVersion`. Will update the `ProductVersion` and `Conditions` of the status to
+/// signal the upgrading / downgrading progress.
+///
+/// # Arguments
+///
+/// * `resource` - The cluster custom resource.
+/// * `current_version` - The current version set in the status `ProductVersion`.
+/// * `target_version` - The target version set in the status `ProductVersion`.
+/// * `spec_version` - The version currently specified in the custom resource.
+/// * `conditions` - The conditions from the custom resource status.
+///
+fn build_version_and_condition<T, V>(
+    resource: &T,
+    current_version: &Option<V>,
+    target_version: &Option<V>,
```

Comment:
Should the be a single ProductVersion<V> Parameter?

Reply:
Done